File: backend/internal/paths/paths.go
// Package paths resolves the canonical OS-standard locations for
// everything the toolkit persists. Mirrors the Rust side's
// crate::paths in overlay/src-tauri so the launcher, the bundled
// backend, and the `rl-toolkit dev` CLI all agree on where state
// lives.
//
// Base directory:
//
//	Linux:   $XDG_DATA_HOME/RLToolkit  (or ~/.local/share/RLToolkit)
//	macOS:   ~/Library/Application Support/RLToolkit
//	Windows: %LOCALAPPDATA%\RLToolkit
//
// Note: Go's stdlib has os.UserConfigDir, os.UserCacheDir, and
// os.UserHomeDir but no equivalent of dirs::data_local_dir(). On
// Windows os.UserConfigDir returns %APPDATA% (roaming), not
// %LOCALAPPDATA%, and on Linux it returns $XDG_CONFIG_HOME, not
// $XDG_DATA_HOME. So we resolve the data-local equivalent by hand to
// stay aligned with Tauri's choice.
package paths

import (
	"errors"
	"os"
	"path/filepath"
	"runtime"
)

// AppDir returns the per-OS application data directory all toolkit
// state lives under.
func AppDir() (string, error) {
	base, err := dataLocalDir()
	if err != nil {
		return "", err
	}
	return filepath.Join(base, "RLToolkit"), nil
}

// DataDir returns <AppDir>/data.
func DataDir() (string, error) {
	app, err := AppDir()
	if err != nil {
		return "", err
	}
	return filepath.Join(app, "data"), nil
}

// PluginsDir returns <AppDir>/plugins.
func PluginsDir() (string, error) {
	app, err := AppDir()
	if err != nil {
		return "", err
	}
	return filepath.Join(app, "plugins"), nil
}

// dataLocalDir mirrors Rust's dirs::data_local_dir():
//
//	Linux:   $XDG_DATA_HOME or $HOME/.local/share
//	macOS:   $HOME/Library/Application Support
//	Windows: %LOCALAPPDATA%
func dataLocalDir() (string, error) {
	switch runtime.GOOS {
	case "windows":
		if v := os.Getenv("LOCALAPPDATA"); v != "" {
			return v, nil
		}
		// Fallback for stripped environments — rare in practice but
		// don't crash on missing env vars.
		home, err := os.UserHomeDir()
		if err != nil {
			return "", err
		}
		return filepath.Join(home, "AppData", "Local"), nil
	case "darwin":
		home, err := os.UserHomeDir()
		if err != nil {
			return "", err
		}
		return filepath.Join(home, "Library", "Application Support"), nil
	case "linux":
		if v := os.Getenv("XDG_DATA_HOME"); v != "" {
			return v, nil
		}
		home, err := os.UserHomeDir()
		if err != nil {
			return "", err
		}
		return filepath.Join(home, ".local", "share"), nil
	default:
		return "", errors.New("unsupported platform")
	}
}

File: overlay/src-tauri/src/paths.rs
//! Canonical filesystem locations for everything the app persists.
//! Mirrors `dirs::data_local_dir()` + `RLToolkit`; the Go side
//! resolves the same path via backend/internal/paths.
//!
//!   Linux:   $XDG_DATA_HOME/RLToolkit (or ~/.local/share/RLToolkit)
//!   macOS:   ~/Library/Application Support/RLToolkit
//!   Windows: %LOCALAPPDATA%\RLToolkit

use std::path::PathBuf;

/// Base dir for all RL Toolkit state. Falls back to `.` for sandboxes
/// where the OS-standard dir can't be resolved.
pub fn base_dir() -> PathBuf {
    dirs::data_local_dir()
        .unwrap_or_else(|| PathBuf::from("."))
        .join("RLToolkit")
}

/// Path to the launcher settings file (overlay enabled, window
/// geometry, user-set plugin/data/RL-API overrides).
pub fn launcher_settings_path() -> PathBuf {
    base_dir().join("launcher.json")
}

/// Default backend data directory. Holds overrides, identity,
/// encounters, datastore, discoveries. The user can override via the
/// Settings dialog; the override persists in launcher.json.
pub fn default_data_dir() -> PathBuf {
    base_dir().join("data")
}

/// Default plugins directory. The user can override via Settings.
pub fn default_plugins_dir() -> PathBuf {
    base_dir().join("plugins")
}

/// Path to the sidecar's stdout/stderr capture file. Catches output
/// that bypasses the Go backend's structured log (runtime panics, raw
/// `fmt.Print`, stack traces).
pub fn sidecar_log_path(date_stamp: &str) -> PathBuf {
    default_data_dir()
        .join("logs")
        .join(format!("sidecar-{date_stamp}.log"))
}

/// Directory holding all rotated log files (launcher-*, backend-*,
/// sidecar-*). Bug-report hand-off: "zip and send this folder."
pub fn logs_dir() -> PathBuf {
    default_data_dir().join("logs")
}
